Add support for Array prop type

sentience opened the pull request:
Add support for props with array types, e.g.

```typescript
  @Prop() listOfStrings: string[] = [];
```

Unfortunately, if the type of the array is not primitive:

```typeScript
  @Prop() listOfStrings: Foo[] = [];

⋮

export type Foo = {
  foo: string;
}
```

Stencil does not give us access to the resolved type of `Foo` in the component metadata. For now, while we continue to investigate how to get what we need out of the Stencil compiler, we treat "unknown" types like this as generic objects, permitting the Elm consumer to set any `Json.Encode.Value` as the value. This also applies to unknown object fields (such as a `handler` field that takes a function.

Closes #4, closes #5, closes #7.

sentience committed:
Add support for array-type props (not working!)

Currently this only works for primitive array types like number[] or
string[]. References to custom types (e.g. Analytics[]) do not get the
type definition expanded by the Stencil compiler, which prevents us from
parsing the type fully.

Need to dig through the internals of Stencil a bit more to see if this
will be possible to support.

sebpearce reviewed (approved):
LGTM. Approved in case you wanted to merge this.

Sad news about the complex types. Maybe we can look at how other Stencil-based component libraries handle those.

sentience commented:
@sebpearce Other component libraries wouldn't have a problem, because they largely don't have to generate their own types in a different type system than TypeScript.

I might still find a way to pull the type information I need out of the the Stencil compiler, but my fallback position will be to make all unsupported props settable as Elm `Json.Encode.Value` types, so that Elm developers can just encode values themselves without type-safety for those particular props.

sentience committed:
Use AnyObjectType as fallback for unknown types

For some reason, the Stencil compiler is refusing to give us the full,
expanded type string for list types. For example:

```
  @Prop() listOfFoo: Foo[]
```

where `Foo` is defined nearby:

```
type Foo = {
  foo: string;
}
```

does not get expanded to `{ foo: string; }[]`, but instead
is described as `Analytics[]` by the compiler, preventing us from
generating a typesafe Elm interface for this type.

While we try to fix Stencil, or otherwise extract the necessary type
information from it, we will treat all such types (including other
unsupported types like functions) as `object` types, which accept any
object that Elm can encode as a `Json.Encode.Value`.

As a part of this change, all prop type classes are now tried for each
type string, rather than using fuzzy regexp matching to find the right
class to use for a given type string. This actually simplifies the code,
while making the matching more robust.

Additionally, the warning message for unrecognised types has been
improved.
